R/convertFlowCore.R

```diff
@@ -0,0 +1,177 @@
+scaleFlowCoreGate <- function (gate, scaleSet) {
+  xChannel <- colnames(gate)[1]
+  yChannel <- colnames(gate)[2]
+  scales <- data.frame(scaleSet$scales)
+  ss <- scales$scale
+  rownames(ss) <- scales$channelName
+
+  scaleX <- ss[xChannel,]
+  scaleY <- ss[yChannel,]
+
+  scaledGate = cbind(applyScale(scaleX, gate[,xChannel]), applyScale(scaleY, gate[,yChannel]))
+  colnames(scaledGate) <- c(xChannel, yChannel)
+  gate[,xChannel] <- unlist(scaledGate[,xChannel])
+  gate[,yChannel] <- unlist(scaledGate[,yChannel])
+
+  return(lapply(seq_len(nrow(gate)), function(i) gate[i,]))
+}
+
+scaleEllipseCenter <- function (center, scaleSet) {
+  xChannel <- names(center)[1]
+  yChannel <- names(center)[2]
+  scales <- data.frame(scaleSet$scales)
+  s <- scales$scale
+  rownames(s) <- scales$channelName
+
+  scaleX <- s[xChannel,]
+  scaleY <- s[yChannel,]
+
+  scaledCenter = cbind(applyScale(scaleX, center[xChannel]), applyScale(scaleY, center[yChannel]))
+  names(scaledCenter) <- c(xChannel, yChannel)
+  center[xChannel] <- unlist(scaledCenter[xChannel])
+  center[yChannel] <- unlist(scaledCenter[yChannel])
+
+  return(center)
+}
+
+#' Convert flowCore to gate
+#'
+#' Converts points in a flowCore gate or flowDensity filter to a CellEngine
+#' Gate and saves it to CellEngine. A ScaleSet must be provided to scale the
+#' gate correctly.
+#'
+#' @param flowObject A flowDensity object or a flowCore Gate
+#' @param scaleSet The CellEngine ScaleSet to be applied.
+#' @export
+#' @examples
+#' \dontrun{
+#' sngl <- flowDensity(file, params)
+#' scaleSet <- getScaleSets("5d2f8b4b21fd0676fb3a6a8c")
+#' convertFromFlowCore(sngl, scaleSet, name = "my gate")
+#' }
+convertFromFlowCore <- function (flowObject, scaleSet, name=NULL) {
+  if (is.null(name)) {
+    name <- flowObject@filterId
+  }
+
+  switch(
+    class(flowObject)[1],
+    "CellPopulation" = { # flowDensity@filter is always a convex polygon
+      flowGate <- flowObject@filter
+      gatePoints <- scaleFlowCoreGate(flowGate, scaleSet)
+      createPolygonGate(
+        scaleSet$experimentId,
+        colnames(flowGate)[1],
+        colnames(flowGate)[2],
+        name,
+        vertices = gatePoints
+      )
+    },
+    "ellipsoidGate" = {
+      params <- covarToParameters(flowObject@cov)
+      x <- flowObject@mean[[1]]
+      y <- flowObject@mean[[2]]
+
+      createEllipseGate(
+        scaleSet$experimentId,
+        colnames(flowObject@cov)[1],
+        colnames(flowObject@cov)[2],
+        name,
+        x,
+        y,
+        params$angle,
+        params$major,
+        params$minor
+      )
+    },
+    "polygonGate" = {
+      flowGate <- flowObject@boundaries
+      gatePoints <- scaleFlowCoreGate(flowGate, scaleSet)
+      createPolygonGate(
+        scaleSet$experimentId,
+        colnames(flowGate)[1],
+        colnames(flowGate)[2],
+        name,
+        vertices = gatePoints
+      )
+    },
+    "rectangleGate" = {
+      createRectangleGate(
+        scaleSet$experimentId,
```

Comment:
I haven't tested yet, but it seems odd that polygons need scaling but rectangles don't.

Reply:
I agree! I haven't really figured out _why_ yet, but simply observed that it appears to be the case. I'll test with some more scales and gates and see if I can puzzle it out.

Reply:
I'm guessing you were working on a linear channel. Rectangle gates are correct, polygon gates are incorrect using this test:

```R
experimentId = "60d1685135401c03803258ad" # shared with you
ss = getScaleSets(experimentId)
rect = cellengine::getGate(experimentId, byName("CD3+")) # ggcyto masks getGate
poly = cellengine::getGate(experimentId, byName("CD3-/CD19-"))
tl=transformList(from=c("Er170Di", "Nd142Di"), list(arcsinhTransform(a=0,b=1/5),arcsinhTransform(a=0,b=1/5)))

file = tempfile()
getEvents(experimentId, byName("LRS046_IL2GMCSF.fcs"), destination=file);
ff = read.FCS(file, transformation=F)

fcrect = convertToFlowCore(rect)
fcpoly = convertToFlowCore(poly)

ggcyto(transform(ff,tl), aes(x=`Er170Di`, y=`Nd142Di`)) +
  ggcyto_par_set(limits=list(x=c(asinh(-5/5), asinh(12000/5)))) +
  geom_hex(bins=56) +
  geom_gate(fcrect) +
  geom_gate(fcpoly)
```
| R | CellEngine |
| --- | --- |
| ![image](https://user-images.githubusercontent.com/469365/122869318-51694b00-d2e9-11eb-8b6a-4092015aed91.png) | ![API testing](https://user-images.githubusercontent.com/469365/122869357-60e89400-d2e9-11eb-8658-ebc809a5aff9.png) |

Good. Now go the other direction:

```R
convertFromFlowCore(fcrect, ss)
convertFromFlowCore(fcpoly, ss)
```
![API testing(1)](https://user-images.githubusercontent.com/469365/122869547-ab6a1080-d2e9-11eb-98cc-0413f07df4d9.png)

```R
# finally
unlink(file)
```

Reply:
Okay, I think this works now